libsupport/src/JSONTracer.cpp

```diff
@@ -0,0 +1,385 @@
+
+  auto usec_ts = std::chrono::duration_cast<std::chrono::microseconds>(
+                     katana::Now().time_since_epoch())
+                     .count();
+  fmt::format_to(
+      std::back_inserter(buf),
+      "\"log\":{{\"msg\":\"{}\",\"timestamp_us\":{},\"max_mem_gb\":{:.3f}"
+      ","
+      "\"mem_gb\":{:.3f},\"arrow_mem_gb\":{:.3f}}}",
+      message, usec_ts, rusage.ru_maxrss / 1024.0 / 1024.0,
+      ParseProcSelfRssBytes() / 1024.0 / 1024.0 / 1024.0,
+      arrow::default_memory_pool()->bytes_allocated() / 1024.0 / 1024.0 /
+          1024.0);
+
+  return fmt::to_string(buf);
+}
+
+std::string
+BuildJSON(
+    const std::string& trace_id, const std::string& span_data,
+    const std::string& log_data, const std::string& tag_data,
+    const std::string& host_data) {
+  fmt::memory_buffer buf;
+  uint32_t host_id = katana::ProgressTracer::GetProgressTracer().GetHostID();
+  static auto begin = katana::Now();
+  auto msec_since_begin = katana::UsSince(begin) / 1000;
+
+  fmt::format_to(
+      std::back_inserter(buf),
+      "{{\"trace_id\":\"{}\",\"host\":{},\"offset_ms\":{},{}", trace_id,
+      host_id, msec_since_begin, span_data);
+  if (!log_data.empty()) {
+    fmt::format_to(std::back_inserter(buf), ",{}", log_data);
+  }
+  if (!tag_data.empty()) {
+    fmt::format_to(std::back_inserter(buf), ",{}", tag_data);
+  }
+  if (!host_data.empty()) {
+    fmt::format_to(std::back_inserter(buf), ",{}", host_data);
+  }
+  fmt::format_to(std::back_inserter(buf), "}}\n");
+  return fmt::to_string(buf);
+}
+
+}  // namespace
+
+std::unique_ptr<katana::JSONTracer>
+katana::JSONTracer::Make(uint32_t host_id, uint32_t num_hosts) {
+  return std::unique_ptr<JSONTracer>(new JSONTracer(host_id, num_hosts));
+}
+
+std::shared_ptr<katana::ProgressSpan>
+katana::JSONTracer::StartSpan(
+    const std::string& span_name,
+    std::shared_ptr<katana::ProgressSpan> child_of) {
+  return JSONSpan::Make(span_name, child_of);
+}
+
+std::shared_ptr<katana::ProgressSpan>
+katana::JSONTracer::StartSpan(
+    const std::string& span_name, const katana::ProgressContext& child_of) {
+  return JSONSpan::Make(span_name, child_of);
+}
+
+std::string
+katana::JSONTracer::Inject(const katana::ProgressContext& ctx) {
+  return ctx.GetTraceID() + "," + ctx.GetSpanID();
+}
+
+std::unique_ptr<katana::ProgressContext>
+katana::JSONTracer::Extract(const std::string& carrier) {
+  size_t split = carrier.find(",");
+  if (split == std::string::npos) {
+    return nullptr;
+  } else {
+    std::string trace_id = carrier.substr(0, split);
+    std::string span_id = carrier.substr(split + 1);
+    return std::unique_ptr<JSONContext>(new JSONContext(trace_id, span_id));
+  }
+}
+
+std::unique_ptr<katana::ProgressContext>
+katana::JSONContext::Clone() const noexcept {
+  return std::unique_ptr<JSONContext>(new JSONContext(trace_id_, span_id_));
+}
+
+katana::JSONSpan::JSONSpan(
+    const std::string& span_name, std::shared_ptr<katana::ProgressSpan> parent)
```

Comment:
why do we take a `shared_ptr<katana::ProgressSpan>` but never do anything with parent itself? We could equally take `ProgressSpan* parent`.

Reply:
It is passed and saved in the parent constructor of ProgressSpan